src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7845,14 +7847,42 @@ private GotoStatementSyntax ParseGotoStatement()
 
         private IfStatementSyntax ParseIfStatement()
         {
-            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.IfKeyword);
-            var @if = this.EatToken(SyntaxKind.IfKeyword);
-            var openParen = this.EatToken(SyntaxKind.OpenParenToken);
-            var condition = this.ParseExpressionCore();
-            var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
-            var statement = this.ParseEmbeddedStatement();
-            var elseClause = ParseElseClauseOpt();
+            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.IfKeyword || this.CurrentToken.Kind == SyntaxKind.ElseKeyword);
+
+            SyntaxToken @if;
+            SyntaxToken openParen;
+            ExpressionSyntax condition;
+            SyntaxToken closeParen;
+            StatementSyntax statement;
+            ElseClauseSyntax elseClause;
 
+            if (this.CurrentToken.Kind == SyntaxKind.IfKeyword)
+            {
+                @if = this.EatToken(SyntaxKind.IfKeyword);
+                openParen = this.EatToken(SyntaxKind.OpenParenToken);
+                condition = this.ParseExpressionCore();
+                closeParen = this.EatToken(SyntaxKind.CloseParenToken);
+                statement = this.ParseEmbeddedStatement();
+                elseClause = this.ParseElseClauseOpt();
+            }
+            else
+            {
+                elseClause = this.ParseElseClauseOpt();
+                elseClause = this.AddErrorToFirstToken(elseClause, ErrorCode.ERR_SyntaxError, SyntaxFacts.GetText(SyntaxKind.IfKeyword));
+
+                // synthesize missing tokens for "if (true) { }":
+                @if = SyntaxToken.CreateMissing(SyntaxKind.IfKeyword, null, null);
```

Comment:
this seems redundant.  You could just always call ``` @if = this.EatToken(SyntaxKind.IfKeyword);``` above.  It will do the right thing if you are/aren't on an 'if' keyword.  This also means you won't need to manually add the 'ERR_SyntaxError' to the 'elseClause'.  

You can do the same for all the rest of the required bits of the else.